vyper/builtin_functions/convert.py

```diff
@@ -1,528 +1,387 @@
+import decimal
+import functools
 import math
-import warnings
-from decimal import Decimal
 
 from vyper import ast as vy_ast
-from vyper.builtin_functions.signatures import signature
 from vyper.codegen.core import (
     LLLnode,
-    add_ofst,
+    bytes_clamp,
+    bytes_data_ptr,
     clamp_basetype,
     get_bytearray_length,
     getpos,
     int_clamp,
-    load_op,
+    load_word,
+    sar,
     shl,
     shr,
-    wordsize,
+    unwrap_location,
 )
+from vyper.codegen.expr import Expr
 from vyper.codegen.types import (
-    DYNAMIC_ARRAY_OVERHEAD,
     BaseType,
+    ByteArrayLike,
     ByteArrayType,
     StringType,
-    get_type,
+    is_base_type,
+    is_bytes_m_type,
+    is_decimal_type,
+    is_integer_type,
 )
-from vyper.evm.opcodes import version_check
-from vyper.exceptions import InvalidLiteral, StructureException, TypeMismatch
-from vyper.utils import DECIMAL_DIVISOR, MemoryPositions, SizeLimits
-
-
-def byte_array_to_num(arg, out_type):
-    """
-    Takes a <32 byte array as input, and outputs a number.
-    """
-    # the location of the bytestring
-    bs_start = (
-        LLLnode.from_list("bs_start", typ=arg.typ, location=arg.location, encoding=arg.encoding)
-        if arg.is_complex_lll
-        else arg
-    )
-
-    if arg.location == "storage":
-        len_ = get_bytearray_length(bs_start)
-        data = LLLnode.from_list(["sload", add_ofst(bs_start, 1)], typ=BaseType("int256"))
-    else:
-        op = load_op(arg.location)
-        len_ = LLLnode.from_list([op, bs_start], typ=BaseType("int256"))
-        data = LLLnode.from_list([op, add_ofst(bs_start, 32)], typ=BaseType("int256"))
-
+from vyper.exceptions import CompilerPanic, InvalidLiteral, StructureException, TypeMismatch
+from vyper.utils import DECIMAL_DIVISOR, SizeLimits, int_bounds
+
+
+def _FAIL(ityp, otyp, pos=None):
+    raise TypeMismatch(f"Can't convert {ityp} to {otyp}", pos)
+
+
+# generate a string representation of a type
+# (makes up for lack of proper hierarchy in LLL type system)
+# (ideally would use type generated during annotation, but
+# not available for builtins)
+def _type_class_of(typ):
+    if is_integer_type(typ):
+        return "int"
+    if is_bytes_m_type(typ):
+        return "bytes_m"
+    if is_decimal_type(typ):
+        return "decimal"
+    if isinstance(typ, BaseType):
+        return typ.typ  # e.g., "bool"
+    if isinstance(typ, ByteArrayType):
+        return "bytes"
+    if isinstance(typ, StringType):
+        return "string"
+
+
+def _input_types(*allowed_types):
+    def decorator(f):
+        @functools.wraps(f)
+        def g(expr, arg, out_typ):
```

Comment:
suggest better names, expr->convert_expr, arg -> to_convert or something

Reply:
out_type -> cast_type or convert_type

Reply:
I agree, `g` is not a great name